IR: Source status, invalidation and Subroutine testing

mlange05 opened the pull request:
This is a preparatory set of changes that is intended to pave the way for a string-conservative Fortran backend that renders changes in a minimally invasive way. For this purpose it introduces an explicit concept of "status" for the `Source` object that IR nodes can be equipped with and extends the invalidation of the source object to track the status of a `Source` object as being in of these three states:
* `SourceStatus.VALID` - A node has not been touched since the `Source` object was created by the frontend.
* `SourceStatus.INVALID_NODE` - The node has been directly altered or changed in in such a way that internal properties are directly contradicting the recorded source string.
* `SourceStatus.INVALID_CHILDREN` - The interior properties (eg. condition expression in conditionals) have not been modified, but some child nodes have been altered, thus rendering the recorded string as partially incorrect.

In addition to the new source status and invalidation in `Transformer` objects, we also now record source objects on `docstring`, `spec` and `body` properties on `Subroutine` objects. I've also made the transformer test non-recursive and test source invalidation explicitly in there now.

Importantly, I've also piggy-bagged a clean-up of some of the old `Subroutine` tests, which needlessly use JIT compilation (slow!) and often have redundancies between them. This makes things a little neater and speeds up the test base a little.

reuterbal commented on loki/frontend/source.py:
```diff
@@ -35,13 +59,16 @@ class Source:
         self.lines = lines
         self.string = string
         self.file = file
+        self.status = status if status else SourceStatus.VALID
```
```suggestion
        self.status = status if status is not None else SourceStatus.VALID
```

You might end up having `auto()` assign `0`, which would be falsey and reordering the enum could then change behaviour.

mlange05 replied:
Urgh, good point! Should we not explicitly avoid `0` then and simply prescribe enum values explicitly?

reuterbal replied:
Meh, I should just have read the documentation:
> [auto](https://docs.python.org/3/library/enum.html#enum.auto)
>    Instances are replaced with an appropriate value for Enum members. [StrEnum](https://docs.python.org/3/library/enum.html#enum.StrEnum) defaults to the lower-cased version of the member name, while other Enums default to 1 and increase from there.

reuterbal commented on loki/frontend/source.py:
```diff
@@ -127,6 +156,12 @@ def clone_lines(self, span=None):
             for idx, line in enumerate(self.string.splitlines())
         ]
 
+    def invalidate(self):
+        """
+        Return an invalid copy of this source with ``status=SourceStatus=INVALID_NODE``.
+        """
+        return self.clone(status=SourceStatus.INVALID_NODE)
+
```
This seems only to be used from tests, not in the actual transformer. Is that intentional?

mlange05 replied:
Sort of. It's a neat utility that can freely be used by user-level transformations, when for example a changed clone of a node is used in a dict-based `Transformer` replacement (as in the tests). However, the only core-level automatic invalidation that happens in this PR is based by in-place changes to the status, so it does not use the utility.

reuterbal commented on loki/ir/expr_visitors.py:
```diff
@@ -237,13 +238,27 @@ class ExpressionTransformer(Transformer):
+        else:
+            new = self.expr_mapper(o)
+        # Invalidate `Source` object if we've changed the expression
+        if kwargs.get('source') and not o == new:
```
```suggestion
        if kwargs.get('source') and o != new:
```
?

reuterbal commented on loki/ir/transformer.py:
```diff
@@ -107,10 +98,12 @@ def _rebuild(self, o, children, **args):
-                if any(child_has_no_source) or len(child_has_no_source) != len(flatten(o.children)):
-                    return self._rebuild_without_source(o, children, **args_frozen)
+            # If any child node has been invalidated, mark this node as invalid too
+            if any(isinstance(c, Node) and not is_source_valid(c) for c in flatten(children)):
+                if is_source_valid(args_frozen.get('source')):
```
Might want to reverse the nesting of the conditionals here to save the loop through children

reuterbal reviewed (changes requested):
Many thanks, this looks great! I've left a few minor comments

mlange05 commented on loki/frontend/fparser.py:
```diff
@@ -1799,7 +1799,7 @@ def visit_Subroutine_Subprogram(self, o, **kwargs):
             if not decls:
                 # No other declarations: add it to the end
-                spec.append(return_var_decl)
+                spec.prepend(return_var_decl)
```
Argh, yes, you're right! The whole thing is not safe, as we always add a `Declaration` node to the spec for prefix-defined return types. This, of course, cannot be string neutral, as it upsets the line-count of the stored source, so we first have to stop adding "ghost declarations'. 😬

mlange05 replied:
This was addressed in PR #552. I've now rebased this.

mlange05 commented:
Ok, @reuterbal , sorry for letting this hang for so long. Following the enabling PR-merge, I've now rebased this and addressed the other comments. Please have another look if you have a sec (assuming tests pass 😉 ).

reuterbal reviewed (approved):
Many thanks, this looks great and #552 is a welcome addition sparked by this :)
